atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -257,22 +232,20 @@ private long getApproximateImmutableTimestamp() {
 
     @Override
     public long getImmutableTimestamp() {
-        long ts = timestampService.getFreshTimestamp();
-        return getImmutableTimestampInternal(ts);
+        long immutableTs = timelockService.getImmutableTimestamp();
+        recordImmutableTimestamp(immutableTs);
+        return immutableTs;
     }
 
-    protected long getImmutableTimestampInternal(long ts) {
-        Long minLocked = lockService.getMinLockedInVersionId(lockClient.getClientId());
-        long ret = minLocked == null ? ts : minLocked;
+    private void recordImmutableTimestamp(long immutableTs) {
         long recentTs = recentImmutableTs.get();
-        while (recentTs < ret) {
-            if (recentImmutableTs.compareAndSet(recentTs, ret)) {
+        while (recentTs < immutableTs) {
+            if (recentImmutableTs.compareAndSet(recentTs, immutableTs)) {
                 break;
             } else {
                 recentTs = recentImmutableTs.get();
             }
```

Comment:
it's doing check and set in a loop while the CAS fails; the else is to read the current value.

`AtomicLong` has a built in way of doing this anyway, will switch to that